usrsctplib/netinet/sctp_callout.c

```diff
@@ -151,117 +281,115 @@ sctp_os_timer_start(sctp_os_timer_t *c, uint32_t to_ticks, void (*ftn) (void *),
 		to_ticks = 1;
 
 	c->c_arg = arg;
-	c->c_flags = (SCTP_CALLOUT_ACTIVE | SCTP_CALLOUT_PENDING);
+	c->c_flags &= (~SCTP_CALLOUT_CANCELLED);
+	c->c_flags |= (SCTP_CALLOUT_ACTIVE | SCTP_CALLOUT_PENDING);
 	c->c_func = ftn;
-	c->c_time = ticks + to_ticks;
-	TAILQ_INSERT_TAIL(&SCTP_BASE_INFO(callqueue), c, tqe);
+	c->c_time = target_ticks;
+	sctp_binary_heap_push(timers_queue, &c->heap_node);
+	SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": callout %p is scheduled with ticks %" PRIu32 "\n", 
+		__func__, ticks, c, target_ticks);
 	SCTP_TIMERQ_UNLOCK();
 }
 
+int
+sctp_os_timer_cancel(sctp_os_timer_t* c) {
+	int ret;
+	SCTP_TIMERQ_LOCK();
+	SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": request to cancel callout %p\n",
+		__func__, ticks, c);
+	ret = sctp_os_timer_cancel_impl(c);
+	SCTP_TIMERQ_UNLOCK();
+	return ret;
+}
+
 int
 sctp_os_timer_stop(sctp_os_timer_t *c)
 {
-	int wakeup_cookie;
-
 	SCTP_TIMERQ_LOCK();
-	/*
-	 * Don't attempt to delete a callout that's not on the queue.
-	 */
-	if (!(c->c_flags & SCTP_CALLOUT_PENDING)) {
-		c->c_flags &= ~SCTP_CALLOUT_ACTIVE;
-		if (sctp_os_timer_current != c) {
-			SCTP_TIMERQ_UNLOCK();
-			return (0);
-		} else {
+	SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": request to stop callout %p\n",
+		__func__, ticks, c);
+	sctp_os_timer_cancel_impl(c);
+
+	while (c == sctp_os_timer_current) {
```

Comment:
Note for reviewers:
There is a `sctp_os_timer_current_changed` conditional variable which is fired whenever `sctp_os_timer_current` is changed. The `sctp_os_timer_stop` must not finish until timer to stop is current timer (if it is not self-stop).